rest_framework_bulk/drf3/mixins.py

```diff
@@ -0,0 +1,122 @@
+from __future__ import print_function, unicode_literals
+import traceback
+from rest_framework import status
+from rest_framework.mixins import CreateModelMixin
+from rest_framework.response import Response
+
+
+__all__ = [
+    'BulkCreateModelMixin',
+    'BulkDestroyModelMixin',
+    'BulkUpdateModelMixin',
+]
+
+
+class BulkCreateModelMixin(CreateModelMixin):
+    """
+    Either create a single or many model instances in bulk by using the
+    Serializers ``many=True`` ability from Django REST >= 2.2.5.
+
+    .. note::
+        This mixin uses the same method to create model instances
+        as ``CreateModelMixin`` because both non-bulk and bulk
+        requests will use ``POST`` request method.
+    """
+
+    def create(self, request, *args, **kwargs):
+        bulk = isinstance(request.data, list)
+
+        if not bulk:
+            return super(BulkCreateModelMixin, self).create(request, *args, **kwargs)
+
+        else:
+            serializer = self.get_serializer(data=request.data, many=True)
+            serializer.is_valid(raise_exception=True)
+            self.perform_bulk_create(serializer)
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+    def perform_bulk_create(self, serializer):
+        return self.perform_create(serializer)
+
+
+class BulkUpdateModelMixin(object):
+    """
+    Update model instances in bulk by using the Serializers
+    ``many=True`` ability from Django REST >= 2.2.5.
+    """
+
+    def get_object(self):
+        try:
+            super(BulkUpdateModelMixin, self).get_object()
+        except AssertionError:
+            # probably happened when called get_object() within options()
+            # via self.metadata_class which is not allowed on list viewset
+            # however since we are enabling PUT here, we should handle the
+            # exception if called within options()
+            # We can simply swallow the exception since that method
+            # does not actually do anything with the returned object
+            for file, line, function, code in traceback.extract_stack():
+                if all((file.endswith('rest_framework/views.py'),
+                        function == 'options')):
```

Comment:
@tomchristie here is the context:

In the super implementation of the [`get_object()`](https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/generics.py#L234-L239), it asserts that `lookup_url_kwarg` is present within `self.kwargs`. That makes sense since `get_object()` is usually called in a URL for a single resource (e.g. `/api/resource/<pk>/`) so `lookup_url_kwarg` will always be present.

The challenge here is that we are trying to add bulk operations with support for `PUT` requests to a list resource which obviously will be lacking a `lookup_url_kwarg`. More particularly the exception occurs when making `OPTIONS` request to the api which uses [`metadata`](https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/views.py#L418) class to get api docs which internally calls [`get_object`](https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/metadata.py#L82).

Here is what is my end goal - detect when metadata is generated in which case ignore the assertion error but in all other cases rerase the exception since that most likely indicates that something is indeed wrong and exception should not be ignored.

What would be the best way of achieving that?

@kevin-brown suggested that we can simply check if `self.request.method == 'OPTIONS'` and then reraise exception but I am not sure if that will catch all the use-cases.